### PR TITLE
feat(frontend): support create index with distributed by syntax

### LINKED ISSUE #5645: RFC: Allow users to specify distribution key for indexes (and tables & MV)
### Background

Consider an index with more than 1 column as its indexed key e.g.

```
create index idx1 on t (k1, k2, k3)
```

In this case we have several choices for distribution keys

1. `(k1, k2, k3)` // use all columns as dist key, cannot do lookup by prefix e.g. `k1 = 2`
2. `(k1, k2)`
3. `(k1)`  // use first columns as dist key, but we are not sure the cardinality of k1 is large enough to make it distribute well.

Currently RisingWave will use 1 and we are considering to move to 3. 

How about allowing users to choose the distribution key?

### Design

Proposed syntax: `distributed by (col1, col2, ...)`

Example:

```
create index idx1 on t (k1, k2, k3) distributed by (k1, k2)
```

Note: According to [RFC: Simplify Concepts of Relational Table](https://singularity-data.quip.com/J0gUAIM6cN38/RFC-Simplify-Concepts-of-Relational-Table), the distribution key is also used as the key of prefix bloom-filter, so the distribution key should be a prefix of the full index keys. Meanwhile, user simply need to specify the distribution key and the key of prefix bloom-filter will also be that.

### Future Optimizations

The idea can also be applied to tables and MViews, and I think there is no reason for not doing so. Of course, the distributed key must be a prefix of the PK of table/MV.

### Discussions

_No response_

### Q&A

_No response_

### PR DESCRIPTION
**Pull request by chenzl25**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

***PLEASE DO NOT LEAVE THIS EMPTY !!!***

Please explain **IN DETAIL** what the changes are in this PR and why they are needed:

- Support user defined distributed key for `create index` statement.
- Index prefix point select can benefit from this PR.

### Example:
```sql
create table t1 (v1 int, v2 int, v3 int);
create index idx1 on t1(v1,v2) include(v3) distributed by(v1);
describe t1;
 Name |                     Type
------+----------------------------------------------
 v1   | Int32
 v2   | Int32
 v3   | Int32
 idx1 | index(v1, v2) include(v3) distributed by(v1)
```

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

If your pull request contains user-facing changes, please specify the types of the changes, and create a release note. Otherwise, please feel free to remove this section.

### Types of user-facing changes

Please keep the types that apply to your changes, and remove those that do not apply.
* SQL commands, functions, and operators

### Release note

Support user defined distributed key for `create index` statement.
User can specify the distribution key (which must be a prefix of index columns) while create an index.
Syntax:
```
CREATE INDEX index_name ON table_name(column_name, [..]) [INCLUDE(column_name, [..])] [DISTRIBUTED BY(column_name)]
```

## Refer to a related PR or issue link (optional)
close #5645

### CODE REVIEW

File: src/frontend/src/handler/create_index.rs [hunk 1]
```diff
@@ -247,7 +269,15 @@ fn assemble_materialize(
...
     PlanRoot::new(
         logical_project,
-        RequiredDist::AnyShard,
+        if distributed_by_columns.is_empty() {
+            // We use the whole index columns as distributed key by default if users
+            // haven't specify the distributed by columns.
+            RequiredDist::PhysicalDist(Distribution::HashShard((0..index_columns.len()).collect()))
+        } else {
+            RequiredDist::PhysicalDist(Distribution::HashShard(
+                (0..distributed_by_columns.len()).collect(),
```
Comment (xiangjinwu) [inline src/frontend/src/handler/create_index.rs:278]: This seems to depend on the `index_columns.starts_with(&distributed_by_columns)` check above, and only the length info is useful after the check. Shall we just pass `distributed_by_columns_len` (alongside `index_columns`) from the caller?

Comment (chenzl25) [reply src/frontend/src/handler/create_index.rs:278]: Sure.

File: src/frontend/src/handler/create_index.rs [hunk 2]
```diff
@@ -87,19 +88,38 @@ pub(crate) fn gen_create_index_plan(
...
     include_columns = include_columns
         .into_iter()
         .filter(|x| set.insert(*x))
         .collect_vec();
 
+    // Remove duplicate columns of distributed by columns
+    let mut set = HashSet::new();
+    distributed_by_columns = distributed_by_columns
+        .into_iter()
+        .filter(|x| set.insert(*x))
```
Comment (BugenZhao) [inline src/frontend/src/handler/create_index.rs:113]:
```suggestion
    let distributed_by_columns = distributed_by_columns
        .into_iter()
        .unique()
```

File: src/frontend/src/optimizer/plan_node/stream_materialize.rs [hunk 3]
```diff
@@ -86,9 +89,11 @@ impl StreamMaterialize {
             Distribution::Single => RequiredDist::single(),
             _ => {
                 if is_index {
-                    RequiredDist::PhysicalDist(Distribution::HashShard(
-                        user_order_by.field_order.iter().map(|x| x.index).collect(),
-                    ))
+                    assert_matches!(
+                        user_distributed_by,
```
Comment (BugenZhao) [inline src/frontend/src/optimizer/plan_node/stream_materialize.rs:93]: Is this field applicable if we're to create a plain materialized view? If not, how about make it an `Option` to imply `is_index`?

File: src/frontend/src/optimizer/plan_node/stream_materialize.rs [hunk 4]
```diff
@@ -73,9 +74,11 @@ impl StreamMaterialize {
     ///
     /// When creating index, `is_index` should be true. Then, materialize will distribute keys
     /// using order by columns, instead of pk.
```
Comment (BugenZhao) [inline src/frontend/src/optimizer/plan_node/stream_materialize.rs:76]: May also update the doc.

**Review by st1page — COMMENTED** [review]
Now adding a planner test for some DDL statements may not be easy #5759 . But maybe we can add some `explain create index` in e2e test as a workaround :thinking: ?

**Review by BugenZhao — COMMENTED** [review]
LGTM

File: src/frontend/src/optimizer/plan_node/stream_materialize.rs (hunk 3 above)
Comment (chenzl25) [reply src/frontend/src/optimizer/plan_node/stream_materialize.rs:93]: It is only used by create index currently, but I think it is also applicable to plain materialized view in the future. How about Adding an assert for plain materialized view? (Assert RequiredDist == Any).

Comment (chenzl25) [reply src/frontend/src/optimizer/plan_node/stream_materialize.rs:93]: It seems all of PlanRoot is created with `required_dist = Any`. I will test it.

**Comment by chenzl25** [comment]
> Now adding a planner test for some DDL statements may not be easy #5759 . But maybe we can add some `explain create index` in e2e test as a workaround 🤔 ?

But e2e test doesn't have automatic tools like `./risedev apply-planner-test`, it is little bit tough for us when explain format is changed.

**Comment by xxchan** [comment]
explain in planner tests is now supported by https://github.com/risingwavelabs/risingwave/pull/5760

**Review by fuyufjh — APPROVED** [review]
LGTM

**Review by BugenZhao — APPROVED** [review]

**Pull request merged**